README
(c) 2014 Thomas Strausbaugh
See below for additional credit where credit is due!

XkeysReportParser. This is a library used to read the key reports from an 
XKeys device. It also allows sending of commands to the XKeys device to 
adjust backlighting, etc. Just sub class the XkeysReportParser class and 
implement the OnKeyUp or OnKeyDown methods to receive the key reports.

Supported models are XK4/8/16 Stick (including Plus/Mobile versions) and 
XK24/60/80/128 (including Plus/Mobile versions).

This class requires the following Arduino library:

https://github.com/felis/USB_Host_Shield_2.0

Oleg Masurov of Circuits at Home also made contributions to this library in
that he made changes to his USB Host Shield Library to facilitate the 
development of this library.

Download the project files and place them in your Arduino libraries
folder. You may need to perform the SPI hardware modification documented by
Circuits at Home for the USB Host Shield depending on whether you are using 
other shields and what they are.

Rick Russell is the author of the library upon which this library is based. 
His original library can be found at:

https://github.com/macoss/XkeysRptParser

This work based on his library was created September 29th, 2014.

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; version 2
of the License.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

/************************************************************************************************
*************************KEY MAPPING FOR XKEYS DEVICES*******************************************

_________________________________________________________________________________________________
|     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |
|  0  |  8  | 16  | 24  | 32  | 40  | 48  | 56  | 64  | 72  | 80  | 88  | 96  | 104 | 112 | 120 |
|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|
|     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |
|  1  |  9  | 17  | 25  | 33  | 41  | 49  | 57  | 65  | 73  | 81  | 89  | 97  | 105 | 113 | 121 |
|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|
|     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |
|  2  | 10  | 18  | 26  | 34  | 42  | 50  | 58  | 66  | 74  | 82  | 90  | 98  | 106 | 114 | 122 |
|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|
|     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |
|  3  | 11  | 19  | 27  | 35  | 43  | 51  | 59  | 67  | 75  | 83  | 91  | 99  | 107 | 115 | 123 |
|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|
|     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |
|  4  | 12  | 20  | 28  | 36  | 44  | 52  | 60  | 68  | 76  | 84  | 92  | 100 | 108 | 116 | 124 |
|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|
|     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |
|  5  | 13  | 21  | 29  | 37  | 45  | 53  | 61  | 69  | 77  | 85  | 93  | 101 | 109 | 117 | 125 |
|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|
|     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |
|  6  | 14  | 22  | 30  | 38  | 46  | 54  | 62  | 70  | 78  | 86  | 94  | 102 | 110 | 118 | 126 |
|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|
|     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |     |
|  7  | 15  | 23  | 31  | 39  | 47  | 55  | 63  | 71  | 79  | 87  | 95  | 103 | 111 | 119 | 127 |
|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|_____|

ORIENT YOUR DEVICE IN THE UPPER LEFT CORNER TO DETERMINE KEY LAYOUT.
EXAMPLE: XK24 DOES NOT HAVE KEY 6, 7 or 32 BUT IT DOES HAVE KEYS 25 THROUGH 29!

XKEYS STICKS ARE NUMBERED SEQUENTIALLY FROM LEFT TO RIGHT STARTING WITH 0.
EXAMPLE: XK-4 STICK HAS KEYS 0, 1, 2 and 3.

*************************************************************************************************
************************************************************************************************/

ChangeLog.txt
October 10th, 2014

 - fixed a bug with red backlighting
 - trimmed SRAM usage
 - added support for XK60/80 and XK128
 - updated README
 - debugging of ATEM example

October 8th, 2014

 - added an ATEM example

October 1st, 2014

 - buried the USB and HID objects into the class to keep them private (makes the library easier to use also)
 - made the OUT report object private
 - added the runLoop method (use in the loop of the sketch)
 - modified examples and keywords file
 - created ChangeLog file
 - general cleanup of code

September 30th, 2014

 - added a new example
 - added keymap to README
 - added reboot method

September 29th, 2014

 - initial commit

hidxkeysrptparser.cpp
// (C) 2014 Thomas Strausbaugh
// See the README for additional credit where credit is due!

// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; version 2
// of the License.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

#include "hidxkeysrptparser.h"

XkeysReportParser::XkeysReportParser() :
Usb(),
Hid(HIDUniversal(&Usb))
{
	for (uint8_t i=0; i<4; i++)
		oldButtons[i] = 0;

  	for (uint8_t i=0; i<RPT_XKEYS_LEN; i++)
		oldRpt[i]	= 0xD;

	for (uint8_t i=0; i<OUT_RPT_LEN; i++)
		rpt[i]	= 0x0;

	_blueIntensity = 128;
	_redIntensity = 128;
	_rowsBlue = 255;
	_rowsRed = 255;
}

void XkeysReportParser::init()
{
  if (Usb.Init() == -1)
            Serial.println(F("OSC did not start."));
            
  delay( 200 );

  if (!Hid.SetReportParser(0, (HIDReportParser*)this))
            ErrorMessage<uint8_t>(PSTR("SetReportParser"), 1  );

  uint32_t setupDelay = millis() + 1000;

  while(millis() < setupDelay) {
    Usb.Task();
  }

}

void XkeysReportParser::Parse(HID *hid, bool is_rpt_id, uint8_t len, uint8_t *buf)
{
	bool match = true;

	// Checking if there are changes in report since the method was last called
	for (uint8_t i=0; i<RPT_XKEYS_LEN; i++)
		if (buf[i] != oldRpt[i])
		{
			match = false;
			break;
		}

	// Calling xkeys event handler
	if (!match)
	{
		for (uint8_t i=0; i<RPT_XKEYS_LEN; i++) oldRpt[i] = buf[i];
	}
	
	uint8_t k;
	if(!STICK) {
		switch (PADSIZE) {
			case 24:
				k = 1;
				break;
			case 60:
				k = 3;
				break;
			case 80:
				k = 3;
				break;
			case 128:
				k = 4;
				break;
		}
	} else {
		k = 1;
	}
	for (uint8_t j = 0; j < k; j++) {

		uint32_t buttons = ((j == 2) && (k == 3)) ? (0x00000000) : (0x00000000 | buf[(j*4)+5]);
		buttons <<= 24;
		buttons |= ((j == 2) && (k == 3)) ? (0x00000000) : ((uint32_t)buf[(j*4)+4] << 16);
		buttons |= ((uint32_t)buf[(j*4)+3] << 8);
		buttons |= ((uint32_t)buf[(j*4)+2] << 0);
		uint32_t changes = (buttons ^ oldButtons[j]);
 

		// Calling Button Event Handler for every button changed
		if (changes) {
			for (uint8_t i=0; i < 32; i++) {
				uint32_t mask = ((uint32_t)0x01 << i);

				if (((mask & changes) > 0)) {
					if ((buttons & mask) > 0) {
						(STICK) ? OnKeyDown(i/8 + 4*(i%8)) : OnKeyDown(i+(j*32));
					} else {
						(STICK) ? OnKeyUp(i/8 + 4*(i%8)) : OnKeyUp(i+(j*32));
                     	     }
				}
			}
			oldButtons[j] = buttons;
		}
	}
}

void XkeysReportParser::runLoop()
{
  Usb.Task();

}

void XkeysReportParser::wipeArray()
{
  for (uint8_t i=0; i<OUT_RPT_LEN; i++)
		rpt[i]	= 0x00;

}

void XkeysReportParser::sendCommand()
{
  uint8_t ret = Hid.SndRpt(OUT_RPT_LEN, rpt);
  Serial.print(F("HID OUT Return: "));
  Serial.println(ret, HEX);
  uint32_t hold = millis() + 25;
  while(millis() < hold)
    Usb.Task();

}

void XkeysReportParser::setBLSteps(uint8_t color, uint8_t *steps)
{
  wipeArray();
  rpt[0] = 172;
  rpt[1] = color;
  for(uint8_t i = 0; i < 10; i++) {
    rpt[i+2] = steps[i];
  }
  sendCommand();
}

void XkeysReportParser::stepBLIntensity(uint8_t color, uint8_t incDec, uint8_t wrap)
{
  wipeArray();
  rpt[0] = 173;
  rpt[1] = color;
  rpt[2] = incDec;
  rpt[3] = wrap;
  sendCommand();
}

void XkeysReportParser::indexSetLEDs(uint8_t led, uint8_t state)
{
  wipeArray();
  rpt[0] = 179;
  rpt[1] = led;
  rpt[2] = state;
  sendCommand();
}

void XkeysReportParser::setFlashFreq(uint8_t freq)
{
  wipeArray();
  rpt[0] = 180;
  rpt[1] = freq;
  sendCommand();
}

void XkeysReportParser::indexSetBL(uint8_t keyId, uint8_t color, uint8_t mode)
{
  wipeArray();
  uint8_t _index;
  rpt[0] = 181;
  if(!STICK) {
    if(color) {
      uint8_t _extra = (PADSIZE == 24) ? 32 : (PADSIZE);
      _index = keyId + _extra;
    } else {
      _index = keyId;
    }
  } else {
    _index = keyId + (2*(keyId/6));
  }
  rpt[1] = _index;
  rpt[2] = mode;
  sendCommand();
}

void XkeysReportParser::setRowBL(uint8_t row, uint8_t color, uint8_t onOff)
{
  wipeArray();
  rpt[0] = 182;
  rpt[1] = color;
  uint8_t _rows;
  if(!STICK) {
    _rows = (color) ? (_rowsRed) : (_rowsBlue);
    bitWrite(_rows, row - 1, onOff);
    (color) ? (_rowsRed = _rows) : (_rowsBlue = _rows);
  } else {
    _rows = (onOff * 255);
  }
  rpt[2] = _rows;
  sendCommand();
}

void XkeysReportParser::toggleBLs()
{
  wipeArray();
  rpt[0] = 184;
  sendCommand();
}

void XkeysReportParser::setLEDs(boolean green, boolean red)
{
  wipeArray();
  uint8_t _LEDs = (((uint8_t)red << 7) | ((uint8_t)green << 6));
  rpt[0] = 186;
  rpt[1] = _LEDs;
  sendCommand();
}

void XkeysReportParser::setBLintensity(uint8_t color, uint8_t intensity)
{
  wipeArray();
  rpt[0] = 187;
  (color) ? (_redIntensity = intensity) : (_blueIntensity = intensity);
  rpt[1] = _blueIntensity;
  rpt[2] = _redIntensity;
  sendCommand();
}

void XkeysReportParser::setUnitID(uint8_t unitID)
{
  wipeArray();
  rpt[0] = 189;
  rpt[1] = unitID;
  sendCommand();
}

void XkeysReportParser::saveBLStateEEPROM(uint8_t save)
{
  wipeArray();
  rpt[0] = 199;
  rpt[1] = save;
  sendCommand();
}

void XkeysReportParser::enableTimeStamp(uint8_t endis)
{
  wipeArray();
  rpt[0] = 210;
  rpt[1] = endis;
  sendCommand();
}

void XkeysReportParser::reboot()
{
  wipeArray();
  rpt[0] = 238;
  sendCommand();
}


hidxkeysrptparser.h
// (C) 2014 Thomas Strausbaugh
// See the README for additional credit where credit is due!

// This program is free software; you can redistribute it and/or
// modify it under the terms of the GNU General Public License
// as published by the Free Software Foundation; version 2
// of the License.

// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.

#if !defined(__HIDXKEYSRPTPARSER_H__)
#define __HIDXKEYSPARSER_H__

#if defined(ARDUINO) && ARDUINO >=100
#include "Arduino.h"
#else
#include <WProgram.h>
#endif

#include "hid.h"
#include "hiduniversal.h"
#include "Usb.h"

/****************************
Supported models are XK 4/8/16 
stick and XK 24/60/80/128.
****************************/

#define RPT_XKEYS_LEN		18
#define OUT_RPT_LEN		35
#define BL_BLUE			0
#define BL_RED			1
#define MODE_OFF		0
#define MODE_ON			1
#define MODE_FLASH		2
#define LED_GREEN		6
#define LED_RED			7

class XkeysReportParser : public HIDReportParser
{
public:

	uint8_t	STICK;
	uint8_t	PADSIZE;

	/**********If the public methods are called from within the class
	(i.e. from the OnKey methods), you do not need to denote them with
	<class_instance>.method. See example.***************************/


	XkeysReportParser();
	void init();
	virtual void Parse(HID *hid, bool is_rpt_id, uint8_t len, uint8_t *buf);
	void runLoop();

	/**********BL stepping methods are only supported on XK128 or on "Plus" or "Mobile" versions of other pads.************/
	void setBLSteps(uint8_t color, uint8_t *steps);
	  // use an array of 10 steps
	  // color: 0 for blue, 1 for red, 255 to reset all to factory default
	  // stepX: 0-255 for each
	void stepBLIntensity(uint8_t color, uint8_t incDec, uint8_t wrap);
	  // color: 0 for blue, 1 for red; incDec: 0 for lower, 1 for higher; wrap: 0 to wrap around, 1 to stop at limit
	/*****************************************************************************/

	void indexSetLEDs(uint8_t led, uint8_t state);
	void setFlashFreq(uint8_t freq); 
        // 0 is fastest, 255 is slowest (approx. 4 seconds)
	void indexSetBL(uint8_t keyId, uint8_t color, uint8_t mode); 
        // Xkeys sticks only have blue backlights--color is disregarded in this case but you must enter something
	void setRowBL(uint8_t row, uint8_t color, uint8_t mode); 
        // Xkeys sticks only have blue backlights--what you pass as color MUST evaluate to 0 and row must evaluate to 1
        // rows are 1 - 8 from top to bottom (entering row 0 will cause an error)
	void toggleBLs();
	  // any BLs that are on will go off. the same ones will come back on. Does not "reverse" the BLs.
	void setLEDs(boolean green, boolean red);
        // true for on and false for off
        // must use indexSetLEDs for flash mode
	void setBLintensity(uint8_t color, uint8_t intensity);
        // 0 is off and 255 is brightest
	void setUnitID(uint8_t unitID);
	  // 0 - 255
	void saveBLStateEEPROM(uint8_t save);
	  // anything other than 0 will save the current BL state as new startup default
	  // ********limited number of times this can be done! Use sparingly!**********
	void enableTimeStamp(uint8_t endis);
	  // 0 to disable, 1 to enable
	void reboot();
	  // supported on "Plus" or "Mobile" versions

protected:
	// Subclass XkeysReportParser and define these methods to receive events
	virtual void OnKeyUp(uint8_t keyId) {}; 
	virtual void OnKeyDown(uint8_t keyId) {};

private:
	USB Usb;
	HIDUniversal Hid;
	void sendCommand();
	void wipeArray();
	uint8_t		_blueIntensity;
	uint8_t		_redIntensity;
	uint8_t		_rowsBlue;
	uint8_t		_rowsRed;
	uint8_t		oldRpt[RPT_XKEYS_LEN];
	uint32_t		oldButtons[4];
	uint8_t		rpt[OUT_RPT_LEN];

};

#endif // __HIDXKEYSRPTPARSER_H__
